pkg/ccl/backupccl/restore_data_processor.go

```diff
@@ -209,23 +222,88 @@ func inputReader(
 	}
 }
 
-func (rd *restoreDataProcessor) runRestoreWorkers(entries chan execinfrapb.RestoreSpanEntry) error {
+type mergedSST struct {
+	entry   execinfrapb.RestoreSpanEntry
+	iter    storage.SimpleMVCCIterator
+	cleanup func()
+}
+
+func (rd *restoreDataProcessor) openSSTs(
+	entry execinfrapb.RestoreSpanEntry, sstCh chan mergedSST,
+) error {
+	ctx := rd.Ctx
+
+	// The sstables only contain MVCC data and no intents, so using an MVCC
+	// iterator is sufficient.
+	var iters []storage.SimpleMVCCIterator
+	var dirs []cloud.ExternalStorage
+
+	// sendIters sends all of the currently accumulated iterators over the
+	// channel.
+	sendIters := func(itersToSend []storage.SimpleMVCCIterator, dirsToSend []cloud.ExternalStorage) {
+		iter := storage.MakeMultiIterator(itersToSend)
+
+		cleanup := func() {
+			for _, dir := range dirsToSend {
+				if err := dir.Close(); err != nil {
+					log.Warningf(ctx, "close export storage failed %v", err)
+				}
+			}
+			for _, iter := range itersToSend {
+				iter.Close()
+			}
+			iter.Close()
+		}
+
+		sstCh <- mergedSST{
+			entry:   entry,
+			iter:    iter,
+			cleanup: cleanup,
+		}
+
+		iters = make([]storage.SimpleMVCCIterator, 0)
+		dirs = make([]cloud.ExternalStorage, 0)
+	}
+
+	log.VEventf(rd.Ctx, 1 /* level */, "ingesting span [%s-%s)", entry.Span.Key, entry.Span.EndKey)
+
+	for _, file := range entry.Files {
+		log.VEventf(ctx, 2, "import file %s which starts at %s", file.Path, entry.Span.Key)
+
+		dir, err := rd.flowCtx.Cfg.ExternalStorage(ctx, file.Dir)
+		if err != nil {
+			return err
+		}
+		// TODO(pbardea): When memory monitoring is added, send the currently
+		// accumulated iterators on the channel if we run into memory pressure.
+		iter, err := storageccl.ExternalSSTReader(ctx, dir, file.Path, rd.spec.Encryption)
+		if err != nil {
+			return err
+		}
+		iters = append(iters, iter)
+	}
+
+	sendIters(iters, dirs)
+	return nil
+}
+
+func (rd *restoreDataProcessor) runRestoreWorkers(ssts chan mergedSST) error {
 	return ctxgroup.GroupWorkers(rd.Ctx, rd.numWorkers, func(ctx context.Context, _ int) error {
 		for {
 			done, err := func() (done bool, _ error) {
-				entry, ok := <-entries
+				sstIter, ok := <-ssts
 				if !ok {
 					done = true
 					return done, nil
 				}
 
-				summary, err := rd.processRestoreSpanEntry(entry)
+				summary, err := rd.processRestoreSpanEntry(sstIter)
 				if err != nil {
 					return done, err
 				}
 
 				select {
-				case rd.progCh <- makeProgressUpdate(summary, entry, rd.spec.PKIDs):
+				case rd.progCh <- makeProgressUpdate(summary, sstIter.entry, rd.spec.PKIDs):
 				case <-ctx.Done():
 					return done, ctx.Err()
```

Comment:
Hm, I think this could mean we exit all the processors while there are still unclosed iters and dirs in the ch?

Reply:
Done (added draining as mentioned above).